### DISCUSSION TITLE
Change Request: Enhancement to the Node API to way to get list of all available rules before running

### DISCUSSION THREAD

**Discussion by stephen-carter-at-sf** [eslint/eslint#18546] [Q&A] [answered]
### ESLint version

v8.57.0

### What problem do you want to solve?

I do not see any way to get a list of the rules that will run dynamically. I am a Lead Software Engineer at Salesforce and I'm helping to develop the next generation of Salesforce Code Analyzer where we basically wrap ESLint as one of our many engines inside of our product. One obstacle that we face is that we'd like to allow the user to preview the available rules from each of our available engines before actually running a scan. 

Today our legacy code for our old ESLint engine uses the deprecated builtinRules constant (or alternatively uses the linter.getRules) to get the base rules of eslint and then iterate over any plugins installed to get their rules. But this is not ideal at all. Having to loop over the plugins to get their rules feels really hacky.

Ideally we'd love to actually just let ESLint (from the Node API) load the config for our users and apply a base config (so we can add some additional plugins) and then ask what rules are available. How can we do this?

### What do you think is the correct solution?

When playing around locally and modifying the eslint code, I noticed that if I add 
```
    getRules() {
        const { cliEngine } = privateMembersMap.get(this);
        const rules = cliEngine.getRules();
        return createRulesMeta(rules);
    }
```
to the ESLint class inside of eslint.js

and then add this type information inside of the index.d.ts file of `@types/eslint` then I can achieve exactly what I want with:

        const baseConfig: Linter.BaseConfig = {
            "plugins": ["@lwc/eslint-plugin-lwc"], // Whatever configuration we want. Just using this as an example
        }
        const options: ESLint.Options = {
            baseConfig: baseConfig
        }
        
        const eslint: ESLint = new ESLint(options); // Should load the users legacy config files automatically.
        // Alternative: const eslint = loadESLint(options); // I would love to just do this but flat config doesn't seem to have an easy way to implement a getRules method.
        
        const ruleMetaData = await eslint.getRules();

Hurray!

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

Additionally, my concern is also for the future. Although this solution idea works with the old config files. It doesn't seem to work with these new flat config files.

Looking at flat-eslint.js for example, I see code that looks like:

```
                /*
                 * All of the plugin and rule information is contained within the
                 * calculated config for the given file.
                 */
                const config = configs.getConfig(filePath);

                if (!config) {
                    throw createExtraneousResultsError();
                }
                const rule = getRuleFromConfig(ruleId, config);
```

So it appears that would be no independent way of getting all the available rules without first saying which file we are interested in running a scan on. Typically our users give us a directory to scan on - and I would hate to have to go file by file in order to get a list of all the rules that were available.

So i'm not sure if the future direction of eslint is in conflict with our need to get a list of available rules from all the base plugins that we apply and all the user plugins that they might apply. Thus i'm here to discuss how we can get this all working.

Side note: We are blocked in upgrading to v9 because it doesn't seem to support eslint-typescript yet (without using an alpha release). So right now we are using v8 still.

**Comment by nzakas** [comment]
> So it appears that would be no independent way of getting all the available rules without first saying which file we are interested in running a scan on.

That is correct.

> So i'm not sure if the future direction of eslint is in conflict with our need to get a list of available rules from all the base plugins that we apply and all the user plugins that they might apply. 

We are rapidly moving towards a world where ESLint doesn't have any core rules, and all rules are provided by plugins, so yeah, it doesn't make sense to add anything into the API to support this feature.

---

With flat config, though, the plugins are all listed directly in the config file, so you can easily import the file and poke at the plugins to get a list of rules that the plugins support. You'd then just have to combine that with the ESLint core rules, which you can currently retrieve like this:

```js
const { builtInRules } = require("eslint/use-at-your-own-risk");
const ruleIds = [...builtInRules.keys()];

// do as you please
```

**Comment by stephen-carter-at-sf** [comment]
@nzakas Thank you for your quick reply. I understand you are moving towards not having any core rules. But in a world with all rules provided by plugins, we would still like to get access to all the plugins that our users have applied so that we can construct a list of all the rules available.

> so you can easily import the file and poke at the plugins to get a list of rules that the plugins support

How? The user has many options as to where his config is... is there no longer a hierarchy of locations and filenames that it can load from? And in the legacy world... it is even worse because I believe config can even live in package.json.

All I see that is available though is eslint.calculateConfigForFile(filePath) which only allows us to get the config based on a specific file path. Is it possible to get all of the configs available so that we can loop over them and get all the plugins, so that we can then get all the rules?

Otherwise, how can we get a list of all the rules that are available?

**Reply by nzakas** [reply, marked as answer]
I'm sorry, I don't have the time to dig into exactly what it is you're doing or how you're configuring things. All I can do is give you some pointers.

1. The config file is just a JavaScript file, so you can `import` it. You'll just need to figure out where the file is to load it.
2. Plugins are [just objects](https://eslint.org/docs/latest/use/configure/plugins) in each config object, so you can poke them directly to see what's there.
3. When you call `calculateConfigForFile()`, you'll get a single object that has a `plugins` key. Once again, you can poke at the plugins there to see which rules they have.

As I said, *there is no way to a list of all available rules*. This isn't a concept that ESLint has and that's not going to change. You're just going to have to figure out a way to piece together the information you want.

**Reply by stephen-carter-at-sf** [reply]
OK, what I'll do for now to get the rules off of the legacy config is take advantage of the fact that getESLintPrivateMembers is exported. So for ESLint 8.57.0 I'll do 

```
function getRuleMapFor(legacyESLint: ESLint) {
    const legacyESLintModule = path.resolve(path.dirname(require.resolve('eslint')),'eslint','eslint.js');
    return require(legacyESLintModule).getESLintPrivateMembers(legacyESLint).cliEngine.getRules();
}
```

and if we upgrade to ESLint 9.x I'll do
```
function getRuleMapFor(legacyESLint: LegacyESLint) {
    const legacyESLintModule = path.resolve(path.dirname(require.resolve('eslint')),'eslint','legacy-eslint.js');
    return require(legacyESLintModule).getESLintPrivateMembers(legacyESLint).cliEngine.getRules();
}
```

That should get me the rules when loadESLint gives me a legacy ESLint. For the new ESLint instances (which use flat config), i'll do as you suggest and walk through the config values.

**Comment by ericmorand** [comment]
@nzakas , about that:

> We are rapidly moving towards a world where ESLint doesn't have any core rules, and all rules are provided by plugins, so yeah, it doesn't make sense to add anything into the API to support this feature.

It makes sense. What is the plugin that provide the core rules of ESLint?

Currently, the documentation doesn't require that a plugin is declared to use the core rules. For example:

```
// eslint.config.js
export default [
    {
        rules: {
            semi: "error"
        }
    }
];
```

https://eslint.org/docs/latest/use/configure/configuration-files

It means that the "semi" rule is available even when no plugin is loaded. I understand that it is a situation that will change but for now this is how it is, plus you won't be able to change this before ESLint 10 without creating a breaking change (or, more accurately, you will have to keep those ambiant rules available even when you move then to a dedicated plugin).

So, today, how can we get the rule modules provided by ESLint itself if `getRules` is not available anymore?

**Reply by stephen-carter-at-sf** [reply]
@nzakas Given that ESlint 8 is no longer supported and our customers are starting to move to ESLint 9. The Salesforce Code Analyzer team would like to also move to ESLint 9 but currently are blocked because we have no way of determining which rules are available for our rule selection mechanisms to work.

Any chance we can get some time with you to discuss our use case more so we can receive guidance from you in this area?

**Reply by nzakas** [reply]
Drop me a line: https://humanwhocodes.com/contact
